src/Models/HydrostaticFreeSurfaceModels/mg_implicit_free_surface_solver.jl

```diff
@@ -78,72 +96,20 @@ function MGImplicitFreeSurfaceSolver(grid::AbstractGrid,
 
-    right_hand_side = Field{Center, Center, Nothing}(grid)
-
     # initialize solver with Δt = nothing so that linear matrix is not computed;
     # see `initialize_matrix` methods
     solver = MultigridSolver(Az_∇h²ᶜᶜᶜ_linear_operation!, ∫ᶻ_Axᶠᶜᶜ, ∫ᶻ_Ayᶜᶠᶜ;
                              template_field = right_hand_side,
+                             matrix,
                              settings...)
 
-    # For updating the diagonal
-    matrix_constructors = constructors(arch, solver.matrix)
-    Nx, Ny = grid.Nx, grid.Ny
-    fill_diag!(matrix_constructors, arch, Nx*Ny, Nx*Ny)
-    diagonal = compute_diag(arch, grid, gravitational_acceleration)
-
-    return MGImplicitFreeSurfaceSolver(solver, vertically_integrated_lateral_areas, placeholder_timestep, right_hand_side, matrix_constructors, diagonal)
-end
-
-"""  
-    fill_diag!(constr, arch, M, N)
-
-We want all elements in the diagonal to be initialized in the sparse matrix encoding in 
-preparation for calling `update_diag!`. `fill_diag!` ensures that 0s are stored in the matrix 
-constructors (rather than not being included as is standard for sparse matrices).
-
-Cannot be easily parallelized as all elements want to update `colptr` and `rowval`.
-"""
-function fill_diag!(constr, arch, M, N)
-    colptr, rowval, nzval = unpack_constructors(arch, constr)
-
-    for i in 1:M
-        col_first = Int(colptr[i])
-        col_last = Int(colptr[i+1] - 1)
-        # Binary search for i in rowval between colfirst and collast
-        search = searchsortedfirst(rowval, i, col_first, col_last, Base.Order.Forward)
-        if search > col_last || rowval[search] != i # Column j does not contain entry A[i,j]
-            nz = colptr[M+1] # the final element of colptr
-            _insert!(rowval, search, i, nz)
-            _insert!(nzval, search, 0, nz)
-            for m in (i + 1):(M + 1)
-                @inbounds colptr[m] += 1
-            end
-        end
-    end
-    constr = constructors(arch, M, N, (colptr, rowval, nzval))
+    return MGImplicitFreeSurfaceSolver(arch, solver, vertically_integrated_lateral_areas, placeholder_timestep, right_hand_side, matrix_constructors, diagonal)
 end
 
+@inline finalize_solver!(solver::MGImplicitFreeSurfaceSolver) = finalize_solver!(solver.multigrid_solver)
 
-"""
-    compute_diag(arch, grid, g)
-
-Construct an `Nx * Ny` array on architecture `arch` with elements `Az / g`,
-where `g` is the gravitational accelaration.
-"""
-function compute_diag(arch, grid, g)
-    diag = arch_array(arch, zeros(eltype(grid), grid.Nx, grid.Ny, 1))
 
-    event_c = launch!(arch, grid, :xy, _compute_diag!, diag, grid, g,
-                      dependencies = device_event(arch))
-    wait(event_c)
-    return diag
-end
-
-@kernel function _compute_diag!(diag, grid, g)
-    i, j = @index(Global, NTuple)
-    @inbounds diag[i, j, 1]  = - Azᶜᶜᶜ(i, j, 1, grid) / g
-end
+finalize_solver!(::Any) = nothing
```

Comment:
actually this `(::Any)` should replace the `finalize_solver!(::MultigridCPUSolver) = nothing` in the `Solvers/multigrid_solver.jl`. Then nothing extra is needed here.